GPU/Vulkan/DrawEngineVulkan.cpp

```diff
@@ -0,0 +1,733 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License 2.0 for more details.
+
+// A copy of the GPL 2.0 should have been included with the program.
+// If not, see http://www.gnu.org/licenses/
+
+// Official git repository and contact information can be found at
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+#include <cassert>
+
+#include "base/logging.h"
+#include "base/timeutil.h"
+#include "math/dataconv.h"
+
+#include "Common/MemoryUtil.h"
+#include "Core/MemMap.h"
+#include "Core/Host.h"
+#include "Core/System.h"
+#include "Core/Reporting.h"
+#include "Core/Config.h"
+#include "Core/CoreTiming.h"
+
+#include "GPU/Math3D.h"
+#include "GPU/GPUState.h"
+#include "GPU/ge_constants.h"
+
+#include "GPU/Common/TextureDecoder.h"
+#include "GPU/Common/SplineCommon.h"
+#include "GPU/Common/TransformCommon.h"
+#include "GPU/Common/VertexDecoderCommon.h"
+#include "GPU/Common/SoftwareTransformCommon.h"
+#include "GPU/Common/DrawEngineCommon.h"
+#include "GPU/Vulkan/DrawEngineVulkan.h"
+#include "GPU/Vulkan/TextureCacheVulkan.h"
+#include "GPU/Vulkan/ShaderManagerVulkan.h"
+#include "GPU/Vulkan/PipelineManagerVulkan.h"
+#include "GPU/Vulkan/FramebufferVulkan.h"
+#include "GPU/Vulkan/GPU_Vulkan.h"
+
+enum {
+	DRAW_BINDING_TEXTURE = 0,
+	DRAW_BINDING_2ND_TEXTURE = 1,
+	DRAW_BINDING_DYNUBO_BASE = 2,
+	DRAW_BINDING_DYNUBO_LIGHT = 3,
+	DRAW_BINDING_DYNUBO_BONE = 4,
+};
+
+enum {
+	TRANSFORMED_VERTEX_BUFFER_SIZE = VERTEX_BUFFER_MAX * sizeof(TransformedVertex)
+};
+
+DrawEngineVulkan::DrawEngineVulkan(VulkanContext *vulkan)
+	:
+	vulkan_(vulkan), 
+	decodedVerts_(0),
+	prevPrim_(GE_PRIM_INVALID),
+	lastVType_(-1),
+	pipelineManager_(nullptr),
+	textureCache_(nullptr),
+	framebufferManager_(nullptr),
+	numDrawCalls(0),
+	vertexCountInDrawCalls(0),
+	decodeCounter_(0),
+	fboTexNeedBind_(false),
+	fboTexBound_(false),
+	curFrame_(0) {
+
+	memset(&decOptions_, 0, sizeof(decOptions_));
+	decOptions_.expandAllUVtoFloat = false;
+	decOptions_.expandAllWeightsToFloat = false;
+	decOptions_.expand8BitNormalsToFloat = false;
+
+	// Allocate nicely aligned memory. Maybe graphics drivers will
+	// appreciate it.
+	// All this is a LOT of memory, need to see if we can cut down somehow.
+	decoded = (u8 *)AllocateMemoryPages(DECODED_VERTEX_BUFFER_SIZE);
+	decIndex = (u16 *)AllocateMemoryPages(DECODED_INDEX_BUFFER_SIZE);
+	splineBuffer = (u8 *)AllocateMemoryPages(SPLINE_BUFFER_SIZE);
+	transformed = (TransformedVertex *)AllocateMemoryPages(TRANSFORMED_VERTEX_BUFFER_SIZE);
+	transformedExpanded = (TransformedVertex *)AllocateMemoryPages(3 * TRANSFORMED_VERTEX_BUFFER_SIZE);
+
+	indexGen.Setup(decIndex);
+
+	// All resources we need for PSP drawing. Usually only bindings 0 and 2-4 are populated.
+	VkDescriptorSetLayoutBinding bindings[5];
```

Comment:
Vulkan is not extended that way for minor versions, instead, additional structs are chained using the pNext member, which is more ABI safe. Changing structs requires a major version upgrade. There's a whole chapter on how Vulkan will be extended in vkspec.pdf :)

But yeah, might as well zero it anyway.